contracts/MatchPayouts.sol

```diff
@@ -0,0 +1,159 @@
+
+  /// @dev When true, the payouts mapping is finalized and cannot be updated
+  bool public finalized;
+
+  /// @dev When true, the contract has been funded and withdrawals can begin. Contract must be
+  /// `finalized` before it can be `funded`
+  bool public funded;
+
+  // =========================================== EVENTS ============================================
+
+  /// @dev Emitted when the `finalized` flag is set
+  event Finalized();
+
+  /// @dev Emitted when the `funded` flag is set
+  event Funded();
+
+  /// @dev Emitted when the funder reclaims the funds in this contract
+  event FundingWithdrawn();
+
+  /// @dev Emitted when a payout `amount` is added to the `recipient`'s payout total
+  event PayoutAdded(address recipient, uint256 amount);
+
+  /// @dev Emitted when a `recipient` withdraws their payout
+  event PayoutClaimed(address recipient);
+
+  // ================================== CONSTRUCTOR AND MODIFIERS ==================================
+
+  /**
+   * @param _owner Address of contract owner
+   * @param _funder Address of funder
+   * @param _dai DAI address
+   */
+  constructor(address _owner, address _funder, IERC20 _dai) {
+    owner = _owner;
+    funder = _funder;
+    dai = _dai;
+  }
+
+  /// @dev Requires caller to be the owner
+  modifier onlyOwner() {
+    require(msg.sender == owner, "MatchPayouts: caller is not the owner");
+    _;
+  }
+
+  /// @dev Only allows method to be called once the payouts mapping has been finalized
+  modifier onlyWhenFinalized() {
+    require(finalized, "MatchPayouts: Payouts not finalized");
+    _;
+  }
+
+  // ======================================= PRIMARY METHODS =======================================
+  // Functions are laid out in the order they will be called over the lifetime of the contract
+
+  /**
+   * @notice Set's the mapping of addresses to their match amount
+   * @dev This will need to be called multiple times to prevent exceeding the block gas limit, based
+   * on the number of grants
+   * @param _payouts Array of `Payout`s to set
+   */
+  function setPayouts(Payout[] calldata _payouts) external onlyOwner {
+    require(!finalized, "MatchPayouts: Payouts already finalized");
+    // Set each payout amount
+    for (uint256 i = 0; i < _payouts.length; i += 1) {
+      payouts[_payouts[i].recipient] = _payouts[i].amount;
+      emit PayoutAdded(_payouts[i].recipient, _payouts[i].amount);
+    }
+  }
+
+  /**
+   * @notice Called by the owner to signal that the payout mapping is finalized
+   * @dev Once called, this cannot be reversed!
+   * @dev We use an explicit method here instead of doing this as part of the `setPayouts()` method
+   * to reduce the chance of accidentally setting this flag
+   */
+  function finalize() external onlyOwner {
+    finalized = true;
+    emit Finalized();
+  }
+
+  /**
+   * @notice Enables funder to withdraw all funds
+   * @dev Escape hatch, intended to be used if the payout mapping is finalized incorrectly. In this
```

Comment:
This says "if the payout mapping is finalized incorrectly" but does not check that it's finalized before allowing the withdraw. Do we want to do that? Or is this comment over-stated?

Reply:
Good point. So the order of operations is:
1. Set payout mapping
2. Finalize payout mapping
3. Funders review, and if they approve, then they fund the contract

So by the time it's funded, it's already finalized, and as a result that check should be unnecessary. We could add that check in, but my thinking was that we trust the multisig (the `funder`), and if something is wrong and funds need to be withdrawn, we don't want to risk blocking that for any reason

Reply:
Just to clarify, the order of steps one and two are enforced by the contract, while the order of step three will be done via coordination with the funders, right? That's fine, I just want to make sure I'm understanding the plan and process. Anyway your explanation makes sense so I think we can leave as is.

Reply:
Yep you are understanding correctly!